Support server-customized presence parameters.

rsashank opened the pull request:
<!-- See README for documentation, or ask in #zulip-terminal if unclear -->
### What does this PR do, and why?

Instead of hard-coded constants, uses server provided values, falling back to the original constants when server version is lower.

### External discussion & connections
<!-- [x] all that apply, specifying topic and adding numbers after # for issues/PRs -->
- [x] Discussed in **#zulip-terminal** in `Support server-customized presence parameters #T1421 #T1463`
- [x] Fully fixes #1421 
- [ ] Partially fixes issue #
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?
<!-- [x] all that apply -->
- [ ] Manually - Behavioral changes
- [ ] Manually - Visual changes
- [x] Adapting existing automated tests
- [x] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [x] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [x] It has a commit summary describing the  motivation and reasoning for the change
- [x] It individually passes linting and tests
- [x] It contains test additions for any new behavior
- [x] It flows clearly from a previous branch commit, and/or prepares for the next commit

neiljp commented:
@rsashank Is this ready for review? I know you have other PRs ongoing, but just wanted to check since there was no label. 

I noticed you handled only one part of the issue so far. For moving this forward, some of my feedback on #1448 may also be helpful.

rsashank commented:
@neiljp, I had tests going on this month, so couldn't get back to this. 

Anyways, I've updated the PR, addressed the other part of the issue. I believe it's ready for review now. Would you suggest removing the global variables `OFFLINE_THRESHOLD_SECS = 140` and `PRESENCE_PING_INTERVAL_SECS = 60` on a whole and just putting the values 140 and 60 in the `.get()` instead when the method is called?

@zulipbot add "PR needs review"

neiljp reviewed (commented):
@rsashank Thanks for the update - I see you added more :+1:

My comments are inline, but except for missing a test and needing an update after rebasing, the PR looks effective and my points are mainly minor - moving, naming, and restructuring, to improve readability, style, maintenance.

While this isn't a huge change, it'd be useful to see this formulated as refactors followed by taking account of the new data from the server. As mentioned in one of the recent community events, this allows each smaller change to be described using different commit messages. Of course one could put all of that text in one commit message, but then it can never be as clear which changes relate to which parts of the code!

rsashank commented:
@zulipbot remove "PR awaiting update"
@zulipbot add "PR needs review"

neiljp commented on tests/model/test_model.py:
```diff
@@ -1440,6 +1442,56 @@ def test__store_typing_duration_settings__with_values(
+            ),
+        ],
+        ids=[
+            "Zulip_2.1_ZFL_None_hard_coded",
```
Nit: ZFL=0 now :)

neiljp commented on zulipterminal/model.py:
```diff
@@ -814,6 +818,25 @@ def _store_typing_duration_settings(self) -> None:
+        before the user should be displayed as offline by clients.
+        This field was added in server version 7.0, ZFL 164.
+        """
+        self.server_offline_threshold = self.initial_data.get(
```
So, `self.server_presence_offline_threshold`?

neiljp reviewed (commented):
@rsashank A few small points left over re the code itself, but otherwise my biggest comment is re the commit structure. Otherwise looks close to merging :+1:

rsashank commented:
@zulipbot remove "PR awaiting update"
@zulipbot add "PR needs review"

rsashank commented on zulipterminal/api_types.py:
```diff
@@ -57,6 +57,14 @@
 TYPING_STOPPED_WAIT_PERIOD: Final = 5000
 TYPING_STARTED_EXPIRY_PERIOD: Final = 15000  # TODO: Needs implementation in ZT
 
+
```
That sounds logical. I've updated the code to reflect the 'move' and show how the new constants are now used in the model. Is this acceptable?

neiljp reviewed (commented):
@rsashank This looks cleaner :+1:

I'll look at this again tomorrow when I'm more awake, otherwise I might have made some of the minor changes below myself and merged:
- The first commit touches two files, so should be reflected in the 'areas' in the commit title
- I'd be inclined to include the `_secs` suffix on the model attributes, since the unit is in the constants and the server data
- The second commit is not just a refactor, it's adding support for a feature, so that shouldn't be in the commit title
- The second commit text has a typo, I'd space the Tests statement by a blank line

In the similar other PR I referenced, the reason for the intermediate commit was for the refactor to use the model attributes, rather than directly the constants themselves. Here that's a very small change, so this is fine. However, perhaps to reiterate the same point as before, generally the smaller you can make refactors that are easy to validate, the quicker and easier it is to review - and merge :) Splitting it up that way can also prompt thoughts on adding tests sooner.

To be clear about the minor points I noted: if you want to make those yourself, just let me know so I don't adjust it myself tomorrow too :)

rsashank commented:
I was just updating this right now; looks like I won't need to push changes from my end anymore 😅. Sorry for any added hassle. It now makes sense; initially, I thought it was a refactor since it wasn't altering functionality per se, but having it customizable directly from the server does seem like added functionality. It's great to have that clear, my bad.

Thanks for the review again!

![image](https://github.com/zulip/zulip-terminal/assets/110327345/a4ef9589-a7ac-413e-b5b0-fd91f89e2bce)

neiljp commented:
@rsashank Thanks - I should have specified within a certain timeframe instead :)

Anyhow, take a look at your branch vs this by fetching the pull request (see in `/tools/`) and use `git range-diff` to compare the branch :)

I'm merging now :tada:
